File: nfc_proof_of_concept/lib/app_layer/file_access/serializers/apdu_response_serializer.dart
import 'dart:typed_data';

import '../../field.dart';
import '../fields/response_fields.dart';


class ApduResponse extends ApduSerializer {
  final ApduData? data;
  final ApduStatusWord statusWord;

  ApduResponse._internal({
    this.data,
    required this.statusWord,
    required String name,
  }) : super(name: name);

  factory ApduResponse.success({Uint8List? data}) {
    return ApduResponse._internal(
      data: (data != null && data.isNotEmpty)
          ? ApduData(data, name: "Response Data")
          : null,
      statusWord: ApduStatusWord.ok,
      name: "Success Response",
    );
  }

  
  factory ApduResponse.error(ApduStatusWord errorStatus) {
    if (errorStatus == ApduStatusWord.ok) {
      throw ArgumentError('Cannot create an error response with SW=9000. Use success() factory.');
    }
    return ApduResponse._internal(
      statusWord: errorStatus,
      name: "Error Response",
    );
  }

  @override
  void setFields() {
    // The R-APDU structure is [Data (optional)] + [Status Word].
    // The base ApduSerializer will correctly handle the case where `data` is null.
    fields = [
      data,
      statusWord,
    ];
  }
}

File: nfc_proof_of_concept/packages/nfc_host_card_emulation/android/src/main/kotlin/io/flutter/plugins/nfc_host_card_emulation/app_layer/file_access/serializers/apdu_response_serializer.kt
package io.flutter.plugins.nfc_host_card_emulation.file_access.serializers

import io.flutter.plugins.nfc_host_card_emulation.app_layer.ApduData
import io.flutter.plugins.nfc_host_card_emulation.app_layer.ApduSerializer
import io.flutter.plugins.nfc_host_card_emulation.app_layer.Bytes
import io.flutter.plugins.nfc_host_card_emulation.file_access.fields.ApduStatusWord

class ApduResponse private constructor(
    val data: ApduData?,
    val statusWord: ApduStatusWord,
    name: String
) : ApduSerializer(name) {

    companion object {
        @JvmStatic
        fun fromBytes(rawResponse: Bytes): ApduResponse {
            require(rawResponse.size >= 2) { "Invalid APDU response: must be at least 2 bytes for the status word. Got ${rawResponse.size} bytes." }

            val dataLength = rawResponse.size - 2
            val responseData = if (dataLength > 0) {
                ApduData("Response Data (Parsed)", rawResponse.sliceArray(0 until dataLength))
            } else {
                null
            }

            val sw1 = rawResponse[dataLength].toInt() and 0xFF
            val sw2 = rawResponse[dataLength + 1].toInt() and 0xFF
            val statusWord = ApduStatusWord.fromBytes(sw1, sw2)

            return ApduResponse(
                data = responseData,
                statusWord = statusWord,
                name = "Parsed Response"
            )
        }

        @JvmStatic
        fun success(data: Bytes? = null): ApduResponse {
            val responseData = if (data != null && data.isNotEmpty()) {
                ApduData("Response Data", data)
            } else {
                null
            }
            return ApduResponse(
                data = responseData,
                statusWord = ApduStatusWord.ok,
                name = "Success Response"
            )
        }

        @JvmStatic
        fun error(errorStatus: ApduStatusWord): ApduResponse {
            require(errorStatus != ApduStatusWord.ok) { "Cannot create an error response with SW=9000. Use success() factory." }
            return ApduResponse(
                data = null,
                statusWord = errorStatus,
                name = "Error Response"
            )
        }
    }

    override fun setFields() {
        fields = listOf(data, statusWord)
    }
}